packages/xchain-haven/src/client.ts

```diff
@@ -0,0 +1,315 @@
+import { Asset, Chain, assetFromString, baseAmount } from '@xchainjs/xchain-util'
+import { convertBip39ToHavenMnemonic } from 'mnemonic-converter'
+
+import { getAssetByTicker } from './assets'
+import { HavenCoreClient } from './haven/haven-core-client'
+import { HavenBalance, HavenTicker, SyncObserver } from './haven/types'
+import { assertIsDefined } from './haven/utils'
+import { HavenClient } from './types/client-types'
+
+class Client extends BaseXChainClient implements XChainClient, HavenClient {
+  private havenSDK: HavenCoreClient
+  private havenMnemonic: string | undefined
+
+  /**
+   * Constructor
+   *
+   * Client has to be initialised with network type and phrase.
+   * It will throw an error if an invalid phrase has been passed.
+   *
+   * @param {XChainClientParams} params
+   *
+   * @throws {"Invalid phrase"} Thrown if the given phase is invalid.
+   */
+  constructor({
+    network = Network.Testnet,
+    rootDerivationPaths = {
+      [Network.Mainnet]: `m/44'/535'/0'/0/`,
+      [Network.Testnet]: `m/44'/535'/0'/0/`,
+      [Network.Stagenet]: `m/44'/535'/0'/0/`,
+    },
+    phrase = '',
+  }: XChainClientParams) {
+    super(Chain.Haven, { network, rootDerivationPaths, phrase })
+    this.havenSDK = new HavenCoreClient()
+    if (this.phrase) {
+      this.initSDK()
+    }
+  }
+
+  async getFees(): Promise<Fees> {
+    // map xchains Fee options to Havens priorites
+    const feePriorities = {
+      [FeeOption.Average]: 1,
+      [FeeOption.Fast]: 2,
+      [FeeOption.Fastest]: 4,
+    }
+
+    const fees: Fees = {} as Fees
+
+    for (const feeOption in feePriorities) {
+      const priority = feePriorities[feeOption as FeeOption]
+      const feeAmount = await this.havenSDK.estimateFees(priority)
+      const feeBaseAmount = baseAmount(feeAmount, 12)
+      fees[feeOption as FeeOption] = feeBaseAmount
+    }
+    fees.type = FeeType.PerByte
+    return fees
+  }
+  getAddress(_walletIndex?: number): string {
+    assertIsDefined(this.phrase)
+    const address = this.havenSDK.getAddress()
+    return address
+  }
+
+  getExplorerUrl(): string {
+    const explorerLink = `https://explorer${
+      this.network === Network.Mainnet ? '' : '-' + this.network
+    }.havenprotocol.org`
+    return explorerLink
+  }
+  getExplorerAddressUrl(_address: string): string {
+    throw new Error('cannot lookup addresses in explorer for haven')
+  }
+  getExplorerTxUrl(txID: string): string {
+    return `${this.getExplorerUrl()}/tx/${txID}`
+  }
+  async getBalance(_address: string, assets?: Asset[]): Promise<Balance[]> {
+    const havenBalance: HavenBalance = await this.havenSDK.getBalance()
+
+    //TODO return all asset balances when no assets param provided?
+    const balances: Balance[] = []
+
+    if (assets) {
+      assets.forEach((asset) => {
+        const assetBalance: Balance = {
+          asset,
+          amount: baseAmount(havenBalance[asset.ticker as HavenTicker].balance, 12),
```

Comment:
Type cast is unsafe and might fail. Do a check if `asset.ticker` is really `HavenTicker` by introducing a type guard in `utils`

```
const isHavenTicker = (ticker: string): ticker is HavenTicker => (_...do the check here..._)
```

Reply:
guard added and in use